modules/file/file.go

```diff
@@ -988,6 +991,49 @@ func followSymLink(link string) (mode os.FileMode, path string, err error) {
 	return
 }
 
+// checks whether the file is compressed or not
+func (f *fileEntry) checkCompressed() bool {
+	var err error
+	f.fd, err = os.Open(f.filename)
+	if err != nil {
+		panic(err)
+	}
+	magic := make([]byte, 2)
+	n, err := f.fd.Read(magic)
+	if err != nil {
+		panic(err)
+	}
+	if n != 2 {
+		return false
+	}
+	_, err = f.fd.Seek(0, 0)
+	if err != nil {
+		panic(err)
+	}
+	if magic[0] == 0x1f && magic[1] == 0x8b {
+		f.compRdr, err = gzip.NewReader(f.fd)
+		if err != nil {
+			panic(err)
+		}
+		return true
+	}
+	return false
+}
+
+// getReader takes a single file and returns the appropriate reader
+// This is done by checking whether the file is compressed or not
+func (f *fileEntry) getReader() io.Reader {
+	var err error
+	if f.checkCompressed() {
+		return f.compRdr
+	}
+	f.fd, err = os.Open(f.filename)
+	if err != nil {
+		panic(err)
+	}
+	return f.fd
+}
```

Comment:
I think the logic for the two functions above needs to be adjusted a bit:
1. in `checkCompressed`, if the file is not compressed, the descriptor is not closed, but another one is opened in `getReader`. It would be better to just reuse the first one.
2. I'm not sure we get much by splitting up `checkCompressed` and `getReader`, especially because `checkCompressed` does a lot more than `checking`. I would merge this two functions together and avoid the file descriptor dance, it's only 40 lines of code anyway.
3. don't `panic` without having an explicit `recover` in that function. In this case, `checkContent` has a recover and would catch the panic, but in MIG it is bad practice to recover outside of the function that panics. If your function needs to return an error, make it so:
```go
func myfunc() (my stuff, err error) {
    defer func() {
        if e:= recover(); e != nil {
            err = fmt.Errorf("myfunc() -> %v", err)
        }
    }()
    panic(err)
}
```